src/core/auth/qgsauthconfigurationstoragedb.cpp

```diff
@@ -0,0 +1,2370 @@
+  {
+    removeMethodConfig( config.id() );
+  }
+
+  QSqlQuery query( authDatabaseConnection() );
+  query.prepare( QStringLiteral( "INSERT INTO %1 (id, name, uri, type, version, config) VALUES (:id, :name, :uri, :type, :version, :config)" ).arg( quotedQualifiedIdentifier( methodConfigTableName() ) ) );
+  query.bindValue( QStringLiteral( ":id" ), config.id() );
+  query.bindValue( QStringLiteral( ":name" ), config.name() );
+  query.bindValue( QStringLiteral( ":uri" ), config.uri() );
+  query.bindValue( QStringLiteral( ":type" ), config.method() );
+  query.bindValue( QStringLiteral( ":version" ), config.version() );
+  query.bindValue( QStringLiteral( ":config" ), payload );
+
+  if ( !authDbQuery( &query ) )
+  {
+    setError( tr( "Failed to store config '%1'" ).arg( config.id() ), Qgis::MessageLevel::Critical );
+    return false;
+  }
+
+  emit methodConfigChanged();
+
+  return true;
+}
+
+bool QgsAuthConfigurationStorageDb::removeMethodConfig( const QString &id )
+{
+  QMutexLocker locker( &mMutex );
+
+  if ( !authDbOpen() )
+  {
+    setError( tr( "Auth db could not be opened" ) );
+    return false;
+  }
+
+  QSqlQuery query( authDatabaseConnection() );
+
+  query.prepare( QStringLiteral( "DELETE FROM %1 WHERE id = :id" ).arg( quotedQualifiedIdentifier( methodConfigTableName() ) ) );
+  query.bindValue( QStringLiteral( ":id" ), id );
+
+  if ( !authDbQuery( &query ) )
+  {
+    setError( tr( "Failed to remove config '%1'" ).arg( id ), Qgis::MessageLevel::Critical );
+    return false;
+  }
+
+  if ( query.numRowsAffected() == 0 )
+  {
+    setError( tr( "Config '%1' does not exist" ).arg( id ), Qgis::MessageLevel::Warning );
+    return false;
+  }
+
+  emit methodConfigChanged();
+
+  return true;
+}
+
+bool QgsAuthConfigurationStorageDb::methodConfigExists( const QString &id ) const
+{
+  QMutexLocker locker( &mMutex );
+
+  if ( !authDbOpen() )
+  {
+    const_cast< QgsAuthConfigurationStorageDb * >( this )->setError( tr( "Auth db could not be opened" ) );
+    return false;
+  }
+
+  QSqlQuery query( authDatabaseConnection() );
+
+  query.prepare( QStringLiteral( "SELECT COUNT( id ) FROM %1 WHERE id = :id" ).arg( quotedQualifiedIdentifier( methodConfigTableName() ) ) );
+  query.bindValue( QStringLiteral( ":id" ), id );
+
+  if ( !authDbQuery( &query ) )
+  {
+    const_cast< QgsAuthConfigurationStorageDb * >( this )->setError( tr( "Failed to query for config '%1'" ).arg( id ), Qgis::MessageLevel::Critical );
+    return false;
+  }
+
+  if ( query.next() )
+  {
+    return query.value( 0 ).toInt() > 0;
+  }
+
+  return false;
+}
+
+bool QgsAuthConfigurationStorageDb::storeAuthSetting( const QString &key, const QString &value )
+{
+  QMutexLocker locker( &mMutex );
+
+  if ( !authDbOpen() )
+  {
+    setError( tr( "Auth db could not be opened" ) );
+    return false;
+  }
+
+  if ( authSettingExists( key ) )
+  {
+    removeAuthSetting( key );
+  }
+
+  QSqlQuery query( authDatabaseConnection() );
+
```

Comment:
More of a design question than a code question -- but I'm curious how this method will work in a multi-user setup? Eg a number of plugins I've come across (and some I maintain) use this method to securely store API tokens after a user authenticates with a service. So I'm wondering what would happen if there's a central auth db setup (say on postgres), and then a plugin attempts to store something inherently user-specific like a API token? Either the db will be locked down and prevent this, or they'll clobber their work colleagues' token. Both situations are bad, and would ultimately break the plugin. :scream:  (Or is the intention here that a centrally managed auth db is  siloed into completely separate user "buckets" via db-level handling? )

Unless this situation is gracefully handled I can see this being a blocker for real-world use cases.

Reply:
Let's make a step back: the original goal was to provide a way to store credentials in a real client-server DB to remove the filesystem sqlite DB which is causing a lot of issues with **QGIS server** deployments in the cloud.

For the **desktop** there is an enterprise use case which I think is interesting (and I spotted in the wild when working with big government orgs in the past): centrally managed shared credentials for company's services.

With the new API you can have read/only storages and you can have multiple storages, so you can imagine using a read/write local storage for user overrides.

Also, you can subclass  QgsAuthConfigurationStorage (if not SIP_ABSTRACT) or QgsAuthConfigurationStorageDb and create your own accessors to the DB with filters for the user.

I thought about adding a 'user' field or something similar to the tables but I thought it was too much specific and it wouldn't ever cover all use cases.

Coming back to your use case (if I get this right): if a plugin uses the auth system for its own secure storage it would probably be capable of managing its own keys naming convention to avoid clashes with other users.